### PR TITLE
graph-view: draft of thread architecture

### PR DESCRIPTION
**Pull request by liam-fitzgerald**
This PR adds two threads for graph-store, create & join. Moving forward, I think we should use threads for semantic actions that operate on several hooks and stores. This has several benefits:
- We can annotate the precise failure cases of the threads. This would address urbit/landscape#70 . You can potentially even imagine trying to `|hi` a ship before you join its group/resource and failing if we cannot `|hi` the ship inside some timeout. 
- Easier juggling of card ordering. For example,  to join an unmanaged chat you must first join its group. In app/chat-view this means we have to encode our current join request onto a wire, send the join card on that wire, and then pick up the rest of the chat joining process in +on-agent after it's finished.
- Rollback logic can be encapsulated inside the threads, allowing these transactions to complete or fail atomically. 
- It makes the stores more accessible from the command line, and better defines the API for the applications.

Downsides:
- We still have to write a `app/graph-view` in order to juggle JSON and vase conversions, but the logic should remain in the threads.
- Threads can get interrupted, but most of the threads should be short-lived enough that this isn't a problem in practice. 

Drafting to get feedback before I proceed further into this architecture:
RFC: @Fang- @belisarius222 @philipcmonk @tacryt-socryp

### CODE REVIEW

**Comment by tacryt-socryp** [comment]
I think this architecture is excellent. I think with JSON conversion / marks for these actions, there will be no need for a `%graph-view` app.

**Comment by joemfb** [comment]
> Easier juggling of card ordering. For example, to join an unmanaged chat you must first join its group. In app/chat-view this means we have to encode our current join request onto a wire, send the join card on that wire, and then pick up the rest of the chat joining process in +on-agent after it's finished.

I just want to note that, in gall agents, these kind of multi-part transactions are best managed be representing them explicitly in the agent state. The wire then becomes a symbolic reference to that state, and the whole operation becomes much more observable, manageable, interruptible, &c. Of course, the complexity scales geometrically with the number of individual operations involved in the multi-part transaction (and the overall number of multi-part transactions the agent implements), so threads become desirable as the numbers go up.

File: pkg/arvo/ted/graph/create.hoon [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+::
+=/  =metadata
+  %*  .  *metadata
+    title  title.action
+    description  description.action
+    date-created  now.bowl
+    creator  our.bowl
+  ==
+=/  act=metadata-action
+  [%add group-path [%graph (en-path:resource rid.action)] metadata]
```
Comment (tacryt-socryp) [inline pkg/arvo/ted/graph/create.hoon:52]: app-name should be passed in as a part of the poke, and will be exactly the same as the existing ones:

`%chat`, `%publish`, `%link`

File: pkg/arvo/lib/graph-view.hoon [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      ==
+    ::
+    ++  groupify  !!
+    ++  invite    !!
+    ++  associated
+      %-  of
+      :~  group+dejs:resource
+          policy+policy:dejs:group-store
+      ==
+
```
Comment (tacryt-socryp) [inline pkg/arvo/lib/graph-view.hoon:59]: blank line

File: pkg/arvo/lib/graph-view.hoon [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    ==
+    ::
+    ++  create
+      %-  ot
+      :~  resource+dejs:resource
+          title+so
+          description+so
+          app+app
+          associated+associated
+      ==
```
Comment (tacryt-socryp) [inline pkg/arvo/lib/graph-view.hoon:33]: `::` in between

File: pkg/arvo/ted/graph/create.hoon [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  =/  m  (strand ,resource)
+  ?:  ?=(%& -.associated)
+    (pure:m p.associated)
+  =/  =action:group-store
+    [%add-group rid p.associated %&]
+  ;<  ~  bind:m  (poke-our %group-store %group-action !>(action))
+  ;<  ~  bind:m
+    (poke-our %group-push-hook %push-hook-action !>([%add rid]))
+  (pure:m rid)
+
```
Comment (tacryt-socryp) [inline pkg/arvo/ted/graph/create.hoon:19]: blank line

File: pkg/arvo/ted/graph/create.hoon [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    date-created  now.bowl
+    creator  our.bowl
+  ==
+=/  act=metadata-action
+  [%add group-path [%graph (en-path:resource rid.action)] metadata]
+;<  ~  bind:m  (poke-our %metadata-hook %metadata-action !>(act))
+;<  ~  bind:m
+  (poke-our %metadata-hook %metadata-hook-action !>([%add-owned group-path]))
+(pure:m !>(~))
+
```
Comment (tacryt-socryp) [inline pkg/arvo/ted/graph/create.hoon:57]: blank line

File: pkg/arvo/ted/graph/delete.hoon [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  |=  rid=resource
+  =/  m  (strand ,(unit resource))
+  ;<  pax=(unit path)  bind:m
+    %+  scry:strandio   ,(unit path)
+    ;:  weld
+      /gx/metadata-store/resource/graph
+      (en-path:resource rid)
+      /noun
+    ==
+  (pure:m (bind pax de-path:resource))
```
Comment (tacryt-socryp) [inline pkg/arvo/ted/graph/delete.hoon:18]: `::` in between arms

File: pkg/arvo/ted/graph/delete.hoon [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  |=  rid=resource
+  =/  m  (strand ,(unit resource))
+  ;<  ugroup=(unit group)  bind:m
+    %+  scry:strandio   ,(unit group)
+    ;:  weld
+      /gx/group-store/resource/graph
+      (en-path:resource rid)
+      /noun
+    ==
+  (pure:m (need ugroup))
```
Comment (tacryt-socryp) [inline pkg/arvo/ted/graph/delete.hoon:29]: `::` in between arms

File: pkg/arvo/ted/graph/delete.hoon [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+++  delete-graph
+  |=  rid=resource
+  =/  m  (strand ,~)
+  ;<  ~  bind:m
+    (poke-our %graph-push-hook %push-hook-action [%remove rid.action])
+  ;<  ~  bind:m
+    (poke-our %graph-store %graph-update [%archive-graph rid.action])
+  (pure:m ~)
+--
+^-  thread:spider
```
Comment (tacryt-socryp) [inline pkg/arvo/ted/graph/delete.hoon:39]: `::` in between cast and core

File: pkg/arvo/ted/graph/delete.hoon [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+?~  ugroup  (fail:m %nonexistent)
+?.  hidden.u.ugroup
+  (delete-graph rid.action)
+;<  ~  bind:m
+  (poke-our %group-push-hook %pull-hook-action [%remove rid.action])
+;<  ~  bind:m
+  (poke-our %group-store %group-action !>([%remove-group rid.action])
+;<  ~  bind:m  (delete-graph rid.action)
+(pure:m !>(~))
+
```
Comment (tacryt-socryp) [inline pkg/arvo/ted/graph/delete.hoon:61]: blank line

File: pkg/arvo/ted/graph/join.hoon [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+  =/  m  (strand ,(unit resource))
+  ;<  pax=(unit path)  bind:m
+    %+  scry:strandio   ,(unit path)
+    ;:  weld
+      /gx/metadata-store/resource/graph
+      (en-path:resource rid)
+      /noun
+    ==
+  (pure:m (bind pax de-path:resource))
+--
```
Comment (tacryt-socryp) [inline pkg/arvo/ted/graph/join.hoon:19]: `::` in between

File: pkg/arvo/ted/graph/join.hoon [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+  %+  poke-our  %group-pull-hook
+  pull-hook-action+!>([%add ship.action rid.action])
+::
+;<  ~  bind:m
+  %+  poke-our  %metadata-hook
+  metadata-hook-action+!>([%add-synced ship.action rid.action])
+;<  ~  bind:m  
+  (poke-our %graph-pull-hook %pull-hook-action !>([%add ship.action rid.action]))
+(pure:m !>(~))
+
```
Comment (tacryt-socryp) [inline pkg/arvo/ted/graph/join.hoon:52]: blank line

File: pkg/arvo/ted/graph/leave.hoon [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+  |=  rid=resource
+  =/  m  (strand ,(unit resource))
+  ;<  pax=(unit path)  bind:m
+    %+  scry:strandio   ,(unit path)
+    ;:  weld
+      /gx/metadata-store/resource/graph
+      (en-path:resource rid)
+      /noun
+    ==
+  (pure:m (bind pax de-path:resource))
```
Comment (tacryt-socryp) [inline pkg/arvo/ted/graph/leave.hoon:18]: `::` in between arm declarations

File: pkg/arvo/ted/graph/leave.hoon [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+  |=  rid=resource
+  =/  m  (strand ,(unit resource))
+  ;<  ugroup=(unit group)  bind:m
+    %+  scry:strandio   ,(unit group)
+    ;:  weld
+      /gx/group-store/resource/graph
+      (en-path:resource rid)
+      /noun
+    ==
+  (pure:m (need ugroup))
```
Comment (tacryt-socryp) [inline pkg/arvo/ted/graph/leave.hoon:29]: `::` in between arm declarations

File: pkg/arvo/ted/graph/leave.hoon [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+++  delete-graph
+  |=  rid=resource
+  =/  m  (strand ,~)
+  ;<  ~  bind:m
+    (poke-our %graph-pull-hook %pull-hook-action [%remove rid.action])
+  ;<  ~  bind:m
+    (poke-our %graph-store %graph-update [%archive-graph rid.action])
+  (pure:m ~)
+--
+^-  thread:spider
```
Comment (tacryt-socryp) [inline pkg/arvo/ted/graph/leave.hoon:39]: `::` in between core and cast

File: pkg/arvo/ted/graph/leave.hoon [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+?~  ugroup  (fail:m %nonexistent)
+?.  hidden.u.ugroup
+  (delete-graph rid.action)
+;<  ~  bind:m
+  (poke-our %group-push-hook %pull-hook-action [%remove rid.action])
+;<  ~  bind:m
+  (poke-our %group-store %group-action !>([%remove-group rid.action])
+;<  ~  bind:m  (delete-graph rid.action)
+(pure:m !>(~))
+
```
Comment (tacryt-socryp) [inline pkg/arvo/ted/graph/leave.hoon:61]: blank line

**Review by tacryt-socryp — COMMENTED** [review]
Looks good. No need to include the hooks.

**Comment by liam-fitzgerald** [comment]
Should be good to go!

**Comment by philipcmonk** [comment]
This all makes sense to me.  One thing to emphasize about threads is that they can hang -- if you really need them to terminate, you need to set a timer.  Honestly, %spider should let you set an optional timeout.  That would simplify %eth-watcher and %glob -- and those are a good template for the feature that should be added to %spider.

Additionally, threads can fail at any time -- for example, if an OTA arrives in the middle that causes %spider to reload, all outstanding threads will be killed.  You absolutely need to make sure that if it fails in the middle it doesn't leave the system in an inconsistent state.  The easiest way to do this is to accumulate all the changes and apply them once at the end, but that's not always possible.

**Comment by liam-fitzgerald** [comment]
superseded by #3479

**Pull request closed without merging**